src/ParallelAlgorithms/Actions/MemoryMonitor/ContributeMemoryData.hpp

```diff
@@ -0,0 +1,191 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <numeric>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Local.hpp"
+#include "Parallel/MemoryMonitor/MemoryMonitor.hpp"
+#include "Parallel/MemoryMonitor/Tags.hpp"
+#include "Parallel/Serialize.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace mem_monitor {
+/*!
+ * \brief Simple action meant to be run on the MemoryMonitor component that
+ * collects sizes from Groups and Nodegroups.
+ *
+ * \details This action collects the sizes of all the local branches of a group
+ * or nodegroup component, computes the total memory usage on a node for each,
+ * then writes it to disk. For groups, the proc with the maximum memory usage is
+ * also reported along with the size on the proc.
+ *
+ * The columns in the dat file for a nodegroup when running on 3 nodes will be
+ *
+ * - %Time
+ * - Size on node 0 (MB)
+ * - Size on node 1 (MB)
+ * - Size on node 2 (MB)
+ * - Average size per node (MB)
+ *
+ * The columns in the dat file for a group when running on 3 nodes will be
+ *
+ * - %Time
+ * - Size on node 0 (MB)
+ * - Size on node 1 (MB)
+ * - Size on node 2 (MB)
+ * - Proc of max size
+ * - Size on proc of max size (MB)
+ * - Average size per node (MB)
+ *
+ * The dat file will be placed in the `/MemoryMonitors/` group in the reduction
+ * file. The name of the dat file is the `pretty_type::name` of the component.
+ */
+template <typename ContributingComponent>
+struct ContributeMemoryData {
```

Comment:
What do you think of changing the name to `ContributeMemoryConsumption` or `MemoryUsed`? Something more specific than `Data`? Feel free to decline or defer to a PR at the end that does a global rename. I realize renames can be rather annoying in the middle of a series of PRs or commits :)

Reply:
I think I'll wait until all the memory monitor stuff is in just so rebasing goes smoother.

How about `ContributeComponentSize`?

Reply:
Yep, both sounds good!